vllm/model_executor/models/deepseek_vl2.py

```diff
@@ -0,0 +1,644 @@
+                                               num_tiles_in_image]
+
+                tile_index += num_tiles_in_image + 1
+
+                # format global and local features
+                # ----------------- global view add newline -----------------
+                # [hw, D] -> [h, w, D]
+                global_features = global_features.view(h, w, n_dim)
+
+                # [D]     -> [h, 1, D]
+                new_lines_in_global = repeat(self.image_newline,
+                                             "d -> h 1 d",
+                                             h=h)
+
+                # cat([h, w, D], [h, 1, D], dim=1) -> [h, w + 1, D]
+                global_features = torch.cat(
+                    [global_features, new_lines_in_global], dim=1)
+
+                # [h, w + 1, D] -> [h * (w + 1), D]
+                global_features = global_features.view(-1, n_dim)
+
+                # ----------------- local view add newline -----------------
+                # [num_height_tiles * num_width_tiles, h * w, D] ->
+                # [num_height_tiles * h, num_width_tiles * w, D]
+                local_features = rearrange(
+                    local_features,
+                    "(th tw) (h w) d -> (th h) (tw w) d",
+                    th=num_height_tiles,
+                    tw=num_width_tiles,
+                    h=h,
+                    w=w)
+
+                # [D] -> [num_height_tiles * h, 1, D]
+                new_lines_in_local = repeat(self.image_newline,
+                                            "d -> (th h) 1 d",
+                                            th=num_height_tiles,
+                                            h=h)
+
+                # [num_height_tiles * h, num_width_tiles * w + 1, D]
+                local_features = torch.cat(
+                    [local_features, new_lines_in_local], dim=1)
+
+                # [num_height_tiles * h, num_width_tiles * w + 1, D]
+                #   --> [(num_height_tiles * h) * (num_width_tiles * w + 1), D]
+                local_features = local_features.view(-1, n_dim)
+
+                # merge global and local tiles
+                if self.global_view_pos == "head":
+                    global_local_features = torch.cat([
+                        global_features,
+                        self.view_seperator[None, :],
+                        local_features,
+                    ],
+                                                      dim=0)
+                else:
+                    global_local_features = torch.cat([
+                        local_features,
+                        self.view_seperator[None, :],
+                        global_features,
+                    ],
+                                                      dim=0)
+
+                images_in_this_batch.append(global_local_features)
+
+            vision_embeddings.extend(images_in_this_batch)
+        return torch.cat(vision_embeddings, dim=0)
```

Comment:
Keep in mind in V1 the return multimodal embeddings need to be `NestedTensor` with its first dimension matching the number of images in the batch.